### PR TITLE
WIP: Pass required non-string arguments at add_/init_ time.

### PR DESCRIPTION
**Pull request by litghost**
Other changes:
 - Add `finish_` calls
 - Fix some pyflakes lint

### CODE REVIEW

File: uxsdcxx/cpp.py [hunk 1]
```diff
@@ -182,13 +218,29 @@ def _gen_load_simple(t: UxsdSimple, input: str) -> str:
...
-def _gen_load_element_complex(t: UxsdElement, parent: str) -> str:
+def _gen_load_element_complex(t: UxsdElement, parent: str, complex_type_map : Dict[str, UxsdComplex]) -> str:
 	assert isinstance(t.type, UxsdComplex)
-	out = ""
+	out = "{\n"
+
+	args = ["data"]
+	load_args = []
+	
+	for attr in complex_type_map[t.type.name].attrs:
```
Comment (duck2) [inline uxsdcxx/cpp.py]: Why not just `t.type.attrs`?

Comment (litghost) [reply uxsdcxx/cpp.py]: Is that the same as what I did?

Comment (duck2) [reply uxsdcxx/cpp.py]: Yes, `t.type` gives the exact same `UxsdComplex` you get from the map.

**Comment by litghost** [comment]
In https://github.com/duck2/uxsdcxx/pull/29/commits/858566b5c8813a4b6d396d2ee68d2b80a95e4ff5 I've added a hand-written example consumer of the new interface, which converts the input (XML in the example) to the capnproto schema.  In the long term if we are using capnproto, this file should be auto-generated, but I wasn't confident when I started how to implement the capnproto output.

I'm going to start on a generator that walks the capnproto message and calls back into the interface.  After that I'll need a VPR binding of the interface to test relative performance under XML and capnproto.  @duck2 Do you have a working VPR binding of the interface?

**Comment by duck2** [comment]
Unfortunately no, I'm working on the writer interface side currently.

The example looks good! I think it illustrates the use case well.

**Comment by duck2** [comment]
Pushed the write interface changes. The base class [looks like this](https://gist.github.com/duck2/b77f2001c4e2fd20bb0695e2955afbd2) now. I think I should start on the VPR interface, since I can test it more easily.

**Comment by litghost** [comment]
> Pushed the write interface changes. The base class [looks like this](https://gist.github.com/duck2/b77f2001c4e2fd20bb0695e2955afbd2) now. I think I should start on the VPR interface, since I can test it more easily.

Can you incorporate my required attribute changes?

**Comment by litghost** [comment]
Oh, and I think the required attribute changes should only apply on the reader interfaces.  The write interfaces are much less important, performance wise.

**Comment by litghost** [comment]
Also I split the interface header out from the XML reading header, which is a good idea.

**Comment by duck2** [comment]
Well, I couldn't find a way to keep the PR open while merging the changes up to now.

**Pull request merged**